### PR TITLE
Feature #72: Support aggregate and analytics functions

### LINKED ISSUE #72: Add support for unsupported Aggregate and Analytic functions
#### Unsupported

- GROUPING[_ID]
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/grouping[_id].htm

- PERCENTILE_CONT
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/percentile_cont.htm

- PERCENTILE_DISC
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/percentile_disc.htm

#### Partly supported without keywords and OVER clause

- AVG
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/avg.htm

- CORR 
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/corr.htm

- COUNT 
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/count.htm

- COVAR_POP
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/covar_pop.htm

- COVAR_SAMP
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/covar_samp.htm

- FIRST_VALUE
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/first_value.htm

- GROUP_CONCAT 
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/group_concat.htm

- LAST_VALUE
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/last_value.htm

- MAX 
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/max.htm

- MEDIAN
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/median.htm

- MIN 
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/min.htm

- REGR_FUNCTIONS
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/regr_function.htm

- STDDEV
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/stddev.htm

- STDDEV_POP
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/stddev_pop.htm

- STDDEV_SAMP
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/stddev_samp.htm

- SUM
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/sum.htm

- VAR_POP
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/var_pop.htm

- VAR_SAMP
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/var_samp.htm

- VARIANCE
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/variance.htm

#### Unsupported

- CUME_DIST
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/cume_dist.htm

- DENSE_RANK
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/dense_rank.htm

- LAG 
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/lag.htm

- LEAD 
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/lead.htm

- NTH_VALUE
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/nth_value.htm

- NTILE
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/ntile.htm

- PERCENT_RANK
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/percent_rank.htm

- PERCENTILE_CONT
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/percentile_cont.htm

- PERCENTILE_DISC
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/percentile_disc.htm

- RANK
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/rank.htm

- RATIO_TO_REPORT
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/ratio_to_report.htm

- ROW_NUMBER
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/row_number.htm

#### Partly supported without keywords and OVER clause

- ANY
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/any.htm

- EVERY
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/every.htm

- LISTAGG
https://docs.exasol.com/sql_references/functions/alphabeticallistfunctions/listagg.htm

### Other
* NAMED WINDOW CLAUSE
https://docs.exasol.com/sql_references/functions/analyticfunctions.htm#microcontent5
* over_clause
https://docs.exasol.com/sql_references/functions/analyticfunctions.htm?Highlight=over_clause

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #72

- [x] Add documentation `select.md`
- [x] Update `list_of_supported_exasol_functions.md`
- Requirement tracing (not necessary)

Refactorings/changes not implemented here:
* #126 
* #125 
* #127 
* #128

### CODE REVIEW

File: src/test/java/com/exasol/sql/dql/select/rendering/TestSelectRendering.java [hunk 1]
```diff
@@ -156,24 +165,132 @@ void testSelectCastFunction() {
...
+                        OverClause.of("window1")
+                                .windowFrame(frame -> frame.type(WindowFrameType.ROWS).unit(UnitType.CURRENT_ROW)
+                                        .exclude(WindowFrameExclusionType.NO_OTHERS)),
+                        " OVER(window1 ROWS CURRENT ROW EXCLUDE NO OTHERS)"),
+                arguments(
+                        OverClause.of("window1")
+                                .windowFrame(frame -> frame.type(WindowFrameType.ROWS)
+                                        .unitBetween(column("col1"), UnitType.CURRENT_ROW, column("col2"),
+                                                UnitType.CURRENT_ROW)
+                                        .exclude(WindowFrameExclusionType.CURRENT_ROW)),
```
Comment (kaklakariada) [inline src/test/java/com/exasol/sql/dql/select/rendering/TestSelectRendering.java:278]: Proposed API for creating an OVER clause.

File: doc/changes/changes_4.5.0.md [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
+# Exasol SQL Statement Builder 4.5.0, released 2021-09-??
 
-Code name: Internal refactorings on "More Predicates"
+Code name: Support more Aggregate and Analytic functions
+
+## Summary
+
+This release supports all aggregate and analytics functsions provided by Exasol, e.g. `GROUPING[_ID]`, `PERCENTILE_CONT`, `NTH_VALUE` and many more. See the [ticket](https://github.com/exasol/sql-statement-builder/issues/72) for a complete list.
+
+We also addes support for the keywords `DISTINCT` and `ANY` as well as the [over_clause](https://docs.exasol.com/sql_references/functions/analyticfunctions.htm?Highlight=over_clause) for analytic functions.
```
Comment (jakobbraun) [inline doc/changes/changes_4.5.0.md:9]:
```suggestion
We also added support for the keywords `DISTINCT` and `ANY` as well as the [over_clause](https://docs.exasol.com/sql_references/functions/analyticfunctions.htm?Highlight=over_clause) for analytic functions.
```

File: doc/user_guide/statements/select.md [hunk 3]
```diff
@@ -165,3 +167,29 @@ select.from().table("t");
...
+```
+
+### Analytic functions
+
+#### Keywords `DISTINCT` and `ALL`
+
+You create an analytic function with a keyword `DISTINCT` or `ALL` like this:
+
+```java
+Function function = AnalyticFunction.of(ExasolAnalyticFunction.ANY, Keyword.DISTINCT,
```
Comment (jakobbraun) [inline doc/user_guide/statements/select.md:187]: Maybe we could do something like a builder here too? It's lots of parameters and I did not get immediately what the mean.

File: src/main/java/com/exasol/sql/expression/rendering/OverClauseRenderer.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        append(")");
+    }
+
+    private void appendPartition(final List<ValueExpression> columns) {
+        if ((columns == null) || columns.isEmpty()) {
+            return;
+        }
+        appendKeyword(" PARTITION BY ");
+
```
Comment (jakobbraun) [inline src/main/java/com/exasol/sql/expression/rendering/OverClauseRenderer.java:47]:
```suggestion
```

File: src/main/java/com/exasol/sql/expression/rendering/OverClauseRenderer.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        if (windowFrameClause.getExclusion() != null) {
+            appendKeyword(" EXCLUDE ");
+            appendKeyword(windowFrameClause.getExclusion().getSqlKeyword());
+        }
+    }
+
+    private void renderUnit(final WindowFrameUnitClause unit) {
+        if ((unit.getType() == UnitType.PRECEEDING) || (unit.getType() == UnitType.FOLLOWING)) {
+            if (unit.getExpression() == null) {
+                throw new IllegalStateException(messageBuilder("E-ESB-2").message("Expression is missing.")
```
Comment (jakobbraun) [inline src/main/java/com/exasol/sql/expression/rendering/OverClauseRenderer.java:94]: can you add more information here? I would not know for what the expression is missing.

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/sql/expression/function/exasol/ExasolAggregateFunction.java [hunk 6]
```diff
@@ -5,11 +5,9 @@
...
  * This class is a list of Aggregate Functions that the Exasol database supports.
  *
- * <p>
- * Currently unsupported functions: GROUPING, PERCENTILE_CONT, PERCENTILE_DISC, OVER clause for all aggregate functions
- * and keywords. See <a href="https://github.com/exasol/sql-statement-builder/issues/72"> github issue # 72</a>.
- * </p>
+ * @deprecated Use enum {@link ExasolAnalyticAggregateFunctions}.
  */
+@Deprecated
 public enum ExasolAggregateFunction implements FunctionName {
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply src/main/java/com/exasol/sql/expression/function/exasol/ExasolAggregateFunction.java:11]: It works with the normal `ExasolFunction`. The `AnalyticFunction` class requires `ExasolAnalyticAggregateFunctions`.

[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/sql/expression/function/exasol/ExasolAggregateFunction.java:11]: :+1:

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**